documentation/adr/0021-use-keyword-args.md

```diff
@@ -0,0 +1,73 @@
+[<-previous](./0020-use-c-mex-api.md) |
+next->
+
+# 21 - Use keyword args in Python and MATLAB
+
+Date: 2021-Feb-16
+
+## Status
+
+Proposal
+
+## Context
+
+### Python 
+
+Currently, we don’t have an established syntax for Python. Because we have a clean slate it makes sense to only have a Pythonic syntax using required positional, optional positional, and optional keyword/value pairs
+
+```python
+def function_name(a, b, *args, **kwargs):
+	# body
+```
+
+ e.g. `function_name(a, b, c, d, keyword_1=val_1, keyword_2=val_2, …)`
+
+Python does not natively support flag arguments, and they are not used in the core language. They are supported in a small number of Python libraries, notably `matplotlib`, which was modelled on the MATLAB API, and may be implemented on any  function as optional string arguments with custom handling.
+
+
+### Matlab 
+
+The Horace/Herbert source uses a wide range of API syntax:
+
+- `flagname` and `-flagname` flag arguments
+- true and false flags: `flagname` (`true`) and `noflagname` (`false`)
+- key-value pairs `'keyword', value, ...`
+- positional parameters
+- optional parameters interspersed with positional parameters in some functions
+
+The key-value pair syntax is equivalent to the Pythonic syntax, e.g. `function_name(a, b, c, d, ‘keywordA’, valA, ‘keywordB’, true, …)`.
+
+The mix of `-flagname` and `flagname` introduces complexity into argument parsing - specifically identifying if a string argument `flagname` a flag or a keyword argument?
+
+
+## Decision
+
+The MATLAB APIs will support positional, optional flag and keyword arguments:
+
+- flag arguments will be prefixed with `-`
+- if the same parameter is passed as both a flag and keyword argument an error will be raised if their values are inconsistent
+- flag arguments may be negated, specified as `-noflagname` (`false`) or `-flagname` (`true`). An error will be raised if this prefix would lead to an ambiguity with another flag argument e.g. defining`-nodes` (use nodes) and `-des`
```

Comment:
Prefer only one option? Otherwise you're guaranteed to have a null flag along with more complicated parsing. 
i.e. Define `-nopix` exclusively, no ambiguity, no added complexity, no NOP flag?
Of course this does lead to also having the key-val argument `'nopix', true/false`

Reply:
By which I mean do we want to disallow negation for clarity and to avoid some flags doing nothing, but convincing users that they have some purpose? As mentioned, without special handling, this may require negation of the relevant keyword arg by default which may lead to more confusion.

Reply:
I would agree with this, meaning that the flag value is `false` by default but is switched to true if the user specifies it. So in the `cut` case the flag name would actually be "nopix" rather than "pix" - e.g. users can specify `cut(..., '-nopix')` as they do now, but `cut(..., '-pix')` which currently does nothing would in future yield an error (flag not defined)... 

A quick look shows that the other negative flags in use are `-noplot` and `-nodisplay` but most other flags adhere to the rule that they are default `false` and switched to `true` if specified... except for `dispersion_plot`.